[cephfs] add config doc for exporting CephFS over NFS

ajarr opened the pull request:
... using NFS-Ganesha NFS server.

Signed-off-by: Ramana raja <rraja@redhat.com>

ajarr commented:
@jtlayton , I think I should add the config setting to turn off Ganesha caching, and turn on delegations. Should we also add the info about rados storing config and client recovery data as you did here,
https://review.gerrithub.io/#/c/403825/6/src/config_samples/ceph.conf
?

mattbenjamin commented:
@ajarr that seems appropriate and useful, to me at least

ajarr commented on doc/cephfs/nfs.rst:
```diff
@@ -0,0 +1,157 @@
+-  'nfs-ganesha' and 'nfs-ganesha-ceph' packages (preferably the latest
+   ganesha v2.5 stable or later stable versions)
+
+   +  RPMs and packages for xenial available at http://download.ceph.com/nfs-ganesha/
```
Good point

ajarr commented:
> I worry a little about this information "bitrotting" to some degree. We're tracking documentation within the ceph repo about an application that expect to be layered over it that is being developed separately. I'm not sure this is the right place for it.

@jtlayton, interesting. So you're thinking about the FSAL_CEPH/NFS-Ganesha developers having to update documentation in two places, NFS-Ganesha repo and Ceph repo?  

I think we at least need to have a placeholder in the Ceph docs that points to the appropriate documentation in the Ganesha repo to highlight the NFS-Ganesha over CephFS work.

Maybe I should just get rid of the text under the `NFS-Ganesha configuration` section in this doc, and add a link that points to  `config_samples/ceph.conf` (that you just updated) in the ganesha repo? This way I'm not duplicating documentation, and the NFS-Ganesha/FSAL developers don't need to update docs in both the repos in the future.

What do you think?

batrick commented:
jenkins render docs

batrick commented on doc/cephfs/nfs.rst:
```diff
@@ -0,0 +1,157 @@
+
+EXPORT {} options
+
+- ``Export_ID`` must be a integer value.
```
@jtlayton can you elaborate on the significance of the export ID? Does it need to be unique across a ganesha cluster?

batrick commented on doc/cephfs/nfs.rst:
```diff
@@ -0,0 +1,157 @@
+
+- ``Path`` is the CephFS dir path to export.
+
+- ``Pseudo`` is the pseudo FS path for NFSv4 exports.
```
+ Clients will use this path when mounting Ganesha.

ajarr replied:
Will add it

batrick commented on doc/cephfs/nfs.rst:
```diff
@@ -0,0 +1,157 @@
+EXPORT block and a CEPH block. Following is an example::
+
+    EXPORT
+    {
```
I feel like this minimal configuration should also disable the Ganesha caching. Ganesha caching should be disabled 100% of the time and we should document taht.

renhwztetecs replied:
when we tested, we found that if nfs-ganehsa has cache, it will cause client cache pressure. Using Attr_Expiration_Time=0 does not close nfs-ganesha's cache, Is there any other parameter to turn off the nfs-ganehsa parameter?

ajarr replied:
> I feel like this minimal configuration should also disable the Ganesha caching. Ganesha caching should be disabled 100% of the time and we should document taht.

@batrick understood.

ajarr replied:
@renhwztetecs , did you also try disabling or minimizing Ganesha's dentry, open fd, inode caching as mentioned here,
https://review.gerrithub.io/#/c/403825/6/src/config_samples/ceph.conf
Please see line 53 to 68

renhwztetecs replied:
@ajarr Thank you very much for your hints.

@ajarr @batrick @jtlayton 
In addition, I think that NFS-Ganehsa lacks a forget mechanism like FUSE(calling ceph's Client::ll_forget()) to periodically release the cache mechanism in libcephfs because of the large number of files passing through NFS-Ganehsa. When writing, the "failing to respond to cache pressure" warnings alert will appear on the libcephfs side. See this link: http://tracker.ceph.com/issues/12334
Any suggestions are very grateful?

batrick replied:
@renhwztetecs if you have an example reproducer, please post to that issue.

batrick commented on doc/cephfs/nfs.rst:
```diff
@@ -0,0 +1,157 @@
+
+From the command line::
+
+  mount -t nfs -o nfsvers=4.1,proto=tcp <ganesha-host-name>:/ <mount-point>
```
`<ganesha-host-name>:<ganesha pseudo path>`

ajarr replied:
Good catch!

batrick commented on doc/cephfs/nfs.rst:
```diff
@@ -0,0 +1,157 @@
+Current limitations
+===================
+
+- A Ganesha server daemon cannot export more than one Ceph Filesystem.
```
O.o

really?

renhwztetecs replied:
Can export multiple directories, as follows:


>EXPORT
{
        Export_Id = 11;
        Path = /test1;
        Pseudo = /test1;
        Access_Type = RW;
        Protocols = 3,4;
        Squash = No_Root_Squash;
        FSAL {
                Name = CEPH;
        }
}
EXPORT
{
        Export_Id = 12;
        Path = /test2;
        Pseudo = /test2;
        Access_Type = RW;
        Protocols = 3,4;
        Squash = No_Root_Squash;
        FSAL {
                Name = CEPH;
        }
}

ajarr replied:
@renhwztetecs ,  yes a single Ganesha daemon can export multiple directories within a Ceph file system. Have you been able to export multiple file systems?

renhwztetecs replied:
@ajarr 
Yes, I can export both FSAL_VFS and FASL Ceph(Version: nfs-ganesha-2.5.4)
>EXPORT
{
        Export_Id = 14;
        Path = /test4;
        Pseudo = /test4;
        Access_Type = RW;
        Protocols = 3,4;
        Squash = No_Root_Squash;
        FSAL {
                Name = CEPH;
        }
}
EXPORT
{
        Export_Id = 77;
        Path = /abc;
        Pseudo = /abc;
        Access_Type = RW;
        Protocols = 3,4;
        Squash = No_Root_Squash;
        FSAL {
                Name = VFS;
        }
}

batrick replied:
Let's reword it to say "FSAL_CEPH can only export one Ceph file system although multiple directories in a Ceph file system may be exported."

ajarr replied:
Done

batrick commented:
Documenting each config options is okay with me, even if the documentation is in danger of rot / getting out-of-sync with Ganesha docs.

This document should emphasize **why** we have suggested/required certain options though. e.g. why is necessary to use `Protocols = 4;`? (Not all configs need this treatment of course.)

We also need a section for the RADOS backed exports.

ceph-jenkins commented:
Doc render available at http://docs.ceph.com/ceph-prs/20915/

ajarr commented:
@batrick 

> Documenting each config options is okay with me, even if the documentation is in danger of rot / getting out-of-sync with Ganesha docs.

OK.

> This document should emphasize why we have suggested/required certain options though. e.g. why is necessary to use Protocols = 4;? (Not all configs need this treatment of course.)

We say that "It is preferred to mount the NFS-Ganesha exports using NFSv4.1+ protocols
to get the benefit of sessions"  that's not sufficient?
https://github.com/ceph/ceph/pull/20915/commits/67be0819acbd03d2bdb355a3be5acccad0a4d424#diff-6b6c587a347d8eab90ea6f18e3511bffR144

batrick commented:
> We say that "It is preferred to mount the NFS-Ganesha exports using NFSv4.1+ protocols to get the benefit of sessions" that's not sufficient?
67be081#diff-6b6c587a347d8eab90ea6f18e3511bffR144

That's a little vague but okay for now. WE can improve it later.

jtlayton commented:
> Maybe I should just get rid of the text under the NFS-Ganesha configuration section in this doc, and add a link that points to config_samples/ceph.conf (that you just updated) in the ganesha repo? This way I'm not duplicating documentation, and the NFS-Ganesha/FSAL developers don't need to update docs in both the repos in the future.

I think this doc probably ought to be transplanted into the nfs-ganesha sources. The sample config file there is a starting point, but a more thorough writeup like this could also be added. A small doc inside ceph for this is fine, but really ought to just contain a minimum nfs-ganesha version required for functionality.

Given that the ganesha codebase doesn't live inside ceph, this will be a constant struggle to keep things in sync: What if they change ganesha's config options in the future? Your doc will have to have sections for old and new and there will be an inevitable lag in getting them updated. When they're part of the ganesha sources, then that should mostly get updated as people change the code.

ajarr commented:
> I think this doc probably ought to be transplanted into the nfs-ganesha sources. The sample config file there is a starting point, but a more thorough writeup like this could also be added. A small doc inside ceph for this is fine, but really ought to just contain a minimum nfs-ganesha version required for functionality.

I added more than just minimum nfs-ganesha version. I couldn't find an appropriate place in nfs-ganesha repo to put the details mentioned here. Didn't want to delay this patch further.

> Given that the ganesha codebase doesn't live inside ceph, this will be a constant struggle to keep things in sync: What if they change ganesha's config options in the future? Your doc will have to have sections for old and new and there will be an inevitable lag in getting them updated. When they're part of the ganesha sources, then that should mostly get updated as people change the code.

Added only setup and usage steps in Ceph repo that shouldn't get obsolete. I point to your doc for ganesha config options,
https://github.com/nfs-ganesha/nfs-ganesha/blob/next/src/config_samples/ceph.conf
Hope this works for now.

ajarr commented:
jenkins render doc

batrick commented:
jenkins render docs

ceph-jenkins commented:
Doc render available at http://docs.ceph.com/ceph-prs/20915/
